proxy/coco/src/keystore.rs

```diff
@@ -1,145 +1,124 @@
 >;
 
-impl Keystorage<File> {
-    /// Create a file-backed keystore, suitable for production use.
-    #[must_use = "must use CocoStore to put/get a key"]
-    pub fn file(paths: &paths::Paths, pw: SecUtf8) -> Self {
-        let key_path = paths.keys_dir().join(KEY_PATH);
-        let crypto = Pwhash::new(pw, *crypto::KDF_PARAMS_PROD);
-        Self {
-            store: FileStorage::new(&key_path, crypto),
-        }
-    }
-
-    /// Fetch the [`keys::SecretKey`]
-    ///
-    /// # Errors
-    ///
-    /// If no key is stored in this store (i.e. [`Self::init`] was not called), decryption fails,
-    /// or some IO error occurred.
-    pub fn get(&self) -> Result<keys::SecretKey, Error> {
-        Ok(self.store.get_key().map(|pair| pair.secret_key)?)
+impl FileStore {
+    fn store(&self, passphrase: SecUtf8) -> FileStorage {
+        let key_path = self.paths.keys_dir().join(KEY_PATH);
+        let crypto = Pwhash::new(passphrase, *crypto::KDF_PARAMS_PROD);
+        FileStorage::new(&key_path, crypto)
     }
+}
 
-    /// Attempt to get a [`keys::SecretKey`], otherwise we create one and store it.
-    ///
-    /// # Errors
-    ///
-    /// Fails with [`FileError`]
-    pub fn init(&mut self) -> Result<keys::SecretKey, Error> {
-        match self.store.get_key() {
-            Ok(keypair) => Ok(keypair.secret_key),
-            Err(file::Error::NoSuchKey) => {
+impl KeyStore for FileStore {
+    fn create_key(&self, passphrase: SecUtf8) -> Result<keys::SecretKey, Error> {
+        let mut store = self.store(passphrase);
+        match store.get_key() {
+            Ok(_keypair) => Err(FileError::KeyExists.into()),
+            Err(FileError::NoSuchKey) => {
                 let key = keys::SecretKey::new();
-                self.store.put_key(key)?;
+                store.put_key(key)?;
                 Ok(key)
             },
             Err(err) => Err(err.into()),
         }
     }
+
+    fn unseal(&self, passphrase: SecUtf8) -> Result<keys::SecretKey, Error> {
+        let key_pair = self.store(passphrase).get_key()?;
+        Ok(key_pair.secret_key)
+    }
 }
 
-impl Keystorage<Memory> {
-    /// Create an in-memory keystore, suitable for testing.
-    ///
-    /// A fresh [`keys::SecretKey`] will be generated every time this variant is
-    /// instantiated.
-    ///
-    /// # Note
-    ///
-    /// This is not feature-gated behind `#[cfg(test)]`, because the sibling `api` crate needs to
-    /// be able to access it. Use with extreme caution, and only from `#[cfg(test)]` code!
-    ///
-    /// # Errors
-    ///
-    /// If sealing the generated key fails.
-    pub fn memory(pw: SecUtf8) -> Result<Self, Error> {
-        let mut store = MemoryStorage::new(Pwhash::new(pw, *crypto::KDF_PARAMS_TEST));
-        let key = keys::SecretKey::new();
-        store.put_key(key)?;
+pub fn memory() -> impl KeyStore + Send + Sync {
+    MemoryStore {
+        key_and_passphrase: std::sync::Mutex::new(None),
+    }
+}
+
+struct MemoryStore {
```

Comment:
Why not use the `MemoryStorage` from `radicle-keystore`? The implementation seems somewhat similar

Reply:
Using `MemoryStorage` was my initial idea but that struct does not have the right APIs. Specifically, there is no way unseal a sealed in-memory password store and get information on whether the password is valid. This is due to the fact that `MemoryStorage` requires the password at construction and it can never be changed.

Reply:
Cool, thanks for the explanation :blush: